src/idiomatic/leveraging-the-type-system/typestate-pattern/typestate-generics.md

```diff
@@ -0,0 +1,141 @@
+## Typestate Pattern with Generics
+
+Generics can be used with the typestate pattern to reduce duplication and allow
+shared logic across state variants, while still encoding state transitions in
+the type system.
+
+```rust
```

Comment:
Same feedback as for the serializer example - let's start this section with a slide that has a problem statement, a straightforward solution, and a demonstration of a coding bug that typestate could prevent here. Then show an illustration with the state machine that we're trying to implement with typestate and say that if we implement a separate type for each state like we did before, we would have a lot of types and a lot of code. Then you can reveal this example.

Reply:
That means splitting it up in 2 slides, where the revealing happens because we move form slide A to slide B? Or if not, do you have an example in the book where I can see your desired approach in action?

Reply:
Right, I'm suggesting to split into two slides: first slide to convey to students what it is that we're trying to achieve with this API on the domain level using straightforward but bug-prone implementation of a builder, and second slide with a typestate version of the builder (what you have right now).